From 68b14c67bcbb4fee22b595973fc4ff1ec18133ef Mon Sep 17 00:00:00 2001
From: WeiguangTWK <weiguangtwk@outlook.com>
Date: Fri, 3 Jul 2026 12:42:34 +0800
Subject: [PATCH] Neko LineageOS: Fix up! Crash on en_GB and hu_ZZ locale and
 Accessibility Feature enabled

1781713189.846 10290 13969 13969 E AndroidRuntime: FATAL EXCEPTION: main
1781713189.846 10290 13969 13969 E AndroidRuntime: Process: com.google.inputmethod.nekolos, PID: 13969
1781713189.846 10290 13969 13969 E AndroidRuntime: android.content.res.Resources$NotFoundException: String resource ID #0x0
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.content.res.Resources.getText(Resources.java:480)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.content.res.Resources.getString(Resources.java:573)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.utils.SubtypeLocaleUtils$1.job(SubtypeLocaleUtils.java:229)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.utils.SubtypeLocaleUtils$1.job(SubtypeLocaleUtils.java:226)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.utils.RunInLocale.runInLocale(RunInLocale.java:40)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.utils.SubtypeLocaleUtils.getSubtypeLocaleDisplayNameInternal(SubtypeLocaleUtils.java:232)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.utils.SubtypeLocaleUtils.getReplacementString(SubtypeLocaleUtils.java:265)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.utils.SubtypeLocaleUtils.getSubtypeDisplayNameInternal(SubtypeLocaleUtils.java:286)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.utils.SubtypeLocaleUtils.getSubtypeDisplayNameInSystemLocale(SubtypeLocaleUtils.java:272)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.accessibility.MainKeyboardAccessibilityDelegate.announceKeyboardLanguage(MainKeyboardAccessibilityDelegate.java:122)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.accessibility.MainKeyboardAccessibilityDelegate.setKeyboard(MainKeyboardAccessibilityDelegate.java:91)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.keyboard.MainKeyboardView.setKeyboard(MainKeyboardView.java:407)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.keyboard.KeyboardSwitcher.setKeyboard(KeyboardSwitcher.java:167)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.keyboard.KeyboardSwitcher.setAlphabetKeyboard(KeyboardSwitcher.java:226)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.keyboard.internal.KeyboardState.setAlphabetKeyboard(KeyboardState.java:305)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.keyboard.internal.KeyboardState.onLoadKeyboard(KeyboardState.java:143)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.keyboard.KeyboardSwitcher.loadKeyboard(KeyboardSwitcher.java:138)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.LatinIME.onStartInputViewInternal(LatinIME.java:1077)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.LatinIME$UIHandler.onStartInputView(LatinIME.java:524)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.inputmethod.latin.LatinIME.onStartInputView(LatinIME.java:899)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.inputmethodservice.InputMethodService.startViews(InputMethodService.java:3216)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.inputmethodservice.InputMethodService.showWindow(InputMethodService.java:3083)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.inputmethodservice.InputMethodService$InputMethodImpl.showSoftInput(InputMethodService.java:975)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.inputmethodservice.InputMethodService$InputMethodImpl.showSoftInputWithToken(InputMethodService.java:937)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.inputmethodservice.IInputMethodWrapper.executeMessage(IInputMethodWrapper.java:231)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.internal.os.HandlerCaller$MyHandler.handleMessage(HandlerCaller.java:44)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:132)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.os.Looper.dispatchMessage(Looper.java:333)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:263)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:367)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:9287)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:566)
1781713189.846 10290 13969 13969 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:929)

Change-Id: I2d737ced98524cb7ebe65a55f17a158e56941d6a
---
 java/res/values/cm_strings.xml                |  2 --
 java/res/values/strings.xml                   |  5 +++++
 .../latin/utils/SubtypeLocaleUtils.java       | 21 ++++++++++++++++---
 3 files changed, 23 insertions(+), 5 deletions(-)

diff --git a/java/res/values/cm_strings.xml b/java/res/values/cm_strings.xml
index f161122..eb2b8ce 100644
--- a/java/res/values/cm_strings.xml
+++ b/java/res/values/cm_strings.xml
@@ -25,8 +25,6 @@ are used to. This keyboard does not provide a dictionary, and it is not tied to
 language among those that use the Latin alphabet. This keyboard is laid out in the Bépo
 disposition rather than other common dispositions for Latin languages. [CHAR LIMIT=25] -->
     <string name="subtype_no_language_bepo">Alphabet (Bépo)</string>
-    <string name="subtype_hu_ZZ">Hungarian (QWERTY)</string>
-
     <string name="keyboard_theme_material_follow_system">Material - follow system settings</string>
 
     <!-- Title of the settings for setting keyboard height -->
diff --git a/java/res/values/strings.xml b/java/res/values/strings.xml
index ba3bedf..f449845 100644
--- a/java/res/values/strings.xml
+++ b/java/res/values/strings.xml
@@ -238,6 +238,8 @@
     <string name="subtype_es_US">Spanish (US)</string>
     <!-- Description for Hinglish (https://en.wikipedia.org/wiki/Hinglish) keyboard subtype [CHAR LIMIT=25] -->
     <string name="subtype_hi_ZZ">Hinglish</string>
+    <!-- Description for Hungarian (QWERTY) keyboard subtype [CHAR LIMIT=25] -->
+    <string name="subtype_hu_ZZ">Hungarian (QWERTY)</string>
     <!-- Description for Serbian (Latin) keyboard subtype [CHAR LIMIT=25]
          (Latin) can be an abbreviation to fit in the CHAR LIMIT. -->
     <string name="subtype_sr_ZZ">Serbian (Latin)</string>
@@ -256,6 +258,9 @@
     <!-- Description for Hinglish (https://en.wikipedia.org/wiki/Hinglish) keyboard subtype with explicit keyboard layout [CHAR LIMIT=25]
          This should be identical to subtype_hi_ZZ aside from the trailing (%s). -->
     <string name="subtype_with_layout_hi_ZZ">Hinglish (<xliff:g id="KEYBOARD_LAYOUT" example="QWERTY">%s</xliff:g>)</string>
+    <!-- Description for Hungarian (QWERTY) keyboard subtype with explicit keyboard layout [CHAR LIMIT=25]
+         This should be identical to subtype_hu_ZZ aside from the trailing (%s). -->
+    <string name="subtype_with_layout_hu_ZZ">Hungarian (<xliff:g id="KEYBOARD_LAYOUT" example="QWERTY">%s</xliff:g>)</string>
     <!-- Description for Serbian (Latin) keyboard subtype with explicit keyboard layout [CHAR LIMIT=25]
          This should be identical to subtype_sr_ZZ aside from the trailing (%s). -->
     <string name="subtype_with_layout_sr_ZZ">Serbian (<xliff:g id="KEYBOARD_LAYOUT" example="QWERTY">%s</xliff:g>)</string>
diff --git a/java/src/com/android/inputmethod/latin/utils/SubtypeLocaleUtils.java b/java/src/com/android/inputmethod/latin/utils/SubtypeLocaleUtils.java
index 54a3fc3..96dfa21 100644
--- a/java/src/com/android/inputmethod/latin/utils/SubtypeLocaleUtils.java
+++ b/java/src/com/android/inputmethod/latin/utils/SubtypeLocaleUtils.java
@@ -222,15 +222,30 @@ public final class SubtypeLocaleUtils {
         }
 
         final String displayName;
-        if (exceptionalNameResId != null) {
+        if (exceptionalNameResId != null && exceptionalNameResId != 0) {
             final RunInLocale<String> getExceptionalName = new RunInLocale<String>() {
                 @Override
                 protected String job(final Resources res) {
-                    return res.getString(exceptionalNameResId);
+                    try {
+                        return res.getString(exceptionalNameResId);
+                    } catch (Resources.NotFoundException e) {
+                        Log.w(TAG, "Missing exceptional locale resource: locale=" + localeString
+                                + " resId=" + exceptionalNameResId);
+                        return null;
+                    }
                 }
             };
-            displayName = getExceptionalName.runInLocale(sResources, displayLocale);
+            final String exceptionalDisplayName =
+                    getExceptionalName.runInLocale(sResources, displayLocale);
+            displayName = exceptionalDisplayName != null
+                    ? exceptionalDisplayName
+                    : LocaleUtils.constructLocaleFromString(localeString)
+                            .getDisplayName(displayLocale);
         } else {
+            if (exceptionalNameResId != null) {
+                Log.w(TAG, "Invalid exceptional locale resource id: locale=" + localeString
+                        + " resId=" + exceptionalNameResId);
+            }
             displayName = LocaleUtils.constructLocaleFromString(localeString)
                     .getDisplayName(displayLocale);
         }
-- 
2.43.0

